v3/lints/cabf_ev/lint_ev_orgid_inconsistent_subj_and_ext.go

```diff
@@ -0,0 +1,114 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * Contributed by Adriano Santoni <adriano.santoni@staff.aruba.it>
+ * of ACTALIS S.p.A. (www.actalis.com).
+ */
+
+package cabf_ev
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+
+	"errors"
+	"regexp"
+)
+
+func init() {
+	lint.RegisterCertificateLint(&lint.CertificateLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_ev_orgid_inconsistent_subj_and_ext",
+			Description:   "Checks that the organizationIdentifier Subject attribute and the CABFOrganizationIdentifier extension are consistent",
+			Citation:      "EVGs 9.2.8 and 9.8.2",
+			Source:        lint.CABFEVGuidelines,
+			EffectiveDate: util.CABFEV_Sec9_2_8_Date,
+		},
+		Lint: NewOrgIdInconsistentSubjAndExt,
+	})
+}
+
+// According to EVGs 9.2.8
+type OrganizationIdentifier struct {
+	Scheme    string
+	Country   string
+	State     string
+	Reference string
+}
+
+func ParseOrgId(orgIdString string, orgId *OrganizationIdentifier) error {
+
+	// This is according to the EVG (stricter than ETSI EN 319 412-1)
+	OrgIdPattern := `^[A-Z]{3}[A-Z]{2}(?:\+[A-Z]{2})?\-.+$`
+
+	compiledRegexp, err := regexp.Compile(OrgIdPattern)
+	if err != nil {
+		// This should neve occur, but one never knows....
+		panic(err)
+	}
+
+	if !compiledRegexp.MatchString(orgIdString) {
+		return errors.New("Cannot parse organizationIdentifier: it is probably invalid")
+	}
+
+	orgId.Scheme = orgIdString[0:3]
+	orgId.Country = orgIdString[3:5]
+
+	if orgIdString[5] == '+' {
+		orgId.State = orgIdString[6:8]
+		orgId.Reference = orgIdString[9:]
+	} else {
+		orgId.Reference = orgIdString[6:]
+	}
+
+	return nil
+}
+
+type orgIdInconsistentSubjAndExt struct{}
+
+func NewOrgIdInconsistentSubjAndExt() lint.LintInterface {
+	return &orgIdInconsistentSubjAndExt{}
+}
+
+func (l *orgIdInconsistentSubjAndExt) CheckApplies(c *x509.Certificate) bool {
+	// It is actually mandatory that, if orgId is present, cabfOrgId be present as well,
+	// however this is already checked by another lint
+	return util.IsEV(c.PolicyIdentifiers) && (len(c.Subject.OrganizationIDs) > 0) &&
+		util.IsExtInCert(c, util.CabfExtensionOrganizationIdentifier)
+}
+
+func (l *orgIdInconsistentSubjAndExt) Execute(c *x509.Certificate) *lint.LintResult {
+	// It should be safe to assume there is only one element in OrganizationIDs
+	var orgId OrganizationIdentifier
+	err := ParseOrgId(c.Subject.OrganizationIDs[0], &orgId)
```

Comment:
Please bounds check this as I'm not sure if the x509 package guarantees the presence of at least one org ID. It's not likely, but panics are a nasty surprise and incoming certs are assumed to be relatively wild.

Reply:
Chris, there is already a bounds check in CheckApplies() so I suppose it's no use to redoit here....am I wrong?